### PR TITLE
Added 3D freelook navigation mode

### PR DESCRIPTION
**Pull request by Zylann**
- Triggered by holding RMB
- Can look around in FPS style
- Can move with WASD, up and down with Q and E
- Movement speed accelerates over time
- Can multiply speed with a modifier key to go faster or slower
- Configurable in editor settings and shortcuts

Preview: https://www.youtube.com/watch?v=lmFVNVHTBFk&feature=youtu.be

This PR gives an alternative to the shortcomings found in #8260

### CODE REVIEW

**Comment by tagcup** [comment]
This is similar to the [walk tool in sketchup](https://help.sketchup.com/sv/article/3000145) which is pretty nice I think.

**Comment by n-pigeon** [comment]
Must have for game engine editors :O

+1

File: editor/plugins/spatial_editor_plugin.h [hunk 1]
```diff
@@ -203,7 +205,7 @@ class SpatialEditorViewport : public Control {
 
 	struct Cursor {
 
-		Vector3 cursor_pos;
+		//Vector3 cursor_pos; // TODO Lots of dead code associated with this, delete it?
```
Comment (akien-mga) [inline editor/plugins/spatial_editor_plugin.h]: Yes.

Comment (Zylann) [inline editor/plugins/spatial_editor_plugin.h]: Sure?
Like, also all code from line 730 to 800 won't do anything. Not sure what it's supposed to do...
https://github.com/godotengine/godot/blob/master/editor/plugins/spatial_editor_plugin.cpp#L730

**Comment by Hinsbart** [comment]
With this, object selection via left mouse button does no longer work.

**Comment by Zylann** [comment]
It works normally on my side. I didn't touch anything in relation to the left button.

**Comment by nunodonato** [comment]
just tried it, works great, thanks for this :D

suggestion: using the mouse wheel will still zoom in/out which creates awkward jumps. What about implementing a speed variable, which is increased and decreased with the mouse wheel, like in blender? That would allow to move really fast and then slow for some details. Just one modifier key is not enough

**Comment by nunodonato** [comment]
@Hinsbart I can't select stuff with LMB but I also couldn't before the freelook mode. perhaps #2852 ?

**Comment by novemberist** [comment]
The reason is holding RMB + WASD/QE automatically switches to "move mode" and doesn't return once the buttons are released. To select objects with the LMB again, you have to manually switch back to select mode(q). Maybe there is a better way to handle mode switching automatically?

**Comment by Zylann** [comment]
Ways to fix:
- modify shortcuts
- prevent edit mode change while freelook is active

**Comment by nunodonato** [comment]
I think it makes sense to prevent mode change while freelook is active. same thing to prevent zoom in/out

**Comment by Zylann** [comment]
@nunodonato I was thinking about doing that too for implementing the speed modifier you suggested. Perhaps I will also add a label showing the speed/zoom level too, in a next PR ;)

**Comment by Zylann** [comment]
They actually aren't editor shortcuts, they are hardcoded:
```cpp
				case KEY_Q: _menu_item_pressed(MENU_TOOL_SELECT); break;
				case KEY_W: _menu_item_pressed(MENU_TOOL_MOVE); break;
				case KEY_E: _menu_item_pressed(MENU_TOOL_ROTATE); break;
				case KEY_R: _menu_item_pressed(MENU_TOOL_SCALE); break;
```

**Comment by akien-mga** [comment]
Would be nice to dehardcode them for us AZERTY users ;)

**Comment by Zylann** [comment]
Also they are handled in SpatialEditor, while freelook is handled in one of the 4 SpatialEditorViewports. First idea I have in mind is to write a function `is_any_viewport_using_freelook()`, and bypass the KEY events if true.

**Comment by novemberist** [comment]
Wouldn't it be great to have freelook work with the camera-node preview and have a "lock camera to view" feature in Godot? Well beyond the scope of this PR, I guess, but would be awesome for adjusting camera positions nevertheless.

**Comment by tagcup** [comment]
BTW, I think it'd be nice to be able to move "up" and "down" too. forward/backward and left/right makes sense for moving on a plane, but in editor, most of the time it's more like flying in the space.

**Comment by nunodonato** [comment]
@tagcup you can move up and down using Q and E, like in blender and ue4 :)

**Comment by Zylann** [comment]
btw I'm changing this in the next PR to be a relative up and down rather than the world-space up vector

**Comment by tagcup** [comment]
@nunodonato Oh, sorry. I've missing a lot of stuff today. Thanks for pointing it out!

**Pull request merged**